src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -211,6 +219,15 @@ ucs_status_t uct_tcp_sockcm_ep_send(uct_tcp_sockcm_ep_t *cep)
 {
     ucs_status_t status;
 
+    /* if there is nothing left to write and got here due to an old/irrelevant write
+     * event, return */
+    if (ucs_test_all_flags(cep->state, UCT_TCP_SOCKCM_EP_SERVER_CREATED |
+                                       UCT_TCP_SOCKCM_EP_DATA_SENT)     ||
+        ucs_test_all_flags(cep->state, UCT_TCP_SOCKCM_EP_ON_CLIENT      |
+                                       UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_SENT) ) {
```

Comment:
1. can we replace UCT_TCP_SOCKCM_EP_SERVER_CREATED by UCT_TCP_SOCKCM_EP_ON_SERVER?
2. so it means that on client side, before it calls notify() API, if we have async event, we will not ignore it (enve though we have nothing to do)?

Reply:
1. you mean in this condition? 
why? UCT_TCP_SOCKCM_EP_SERVER_CREATED is 'stronger'.
2. yes. we will ignore events that arrive after we finished sending. if there is a send attempt before notify() API, it can be a valid one - for sending priv_data, a notify message or parts of them.

Reply:
what we really need to check is UCT_TCP_SOCKCM_EP_ON_SERVER. UCT_TCP_SOCKCM_EP_SERVER_CREATED  can be an assertion
2. but after all priv data is sent, and before notify() API is called, we don't have anything to send. how do we "ignore" in this case?